designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,550 @@
+};
+```
+
+### The `@eslint/config` Package
+
+Because some of the operations ESLint currently do are quite complicated, this design includes a utility package called `@eslint/config` that users can use for common tasks. This package contains the following utilities:
+
+1. `Config` - a class that both wraps a configuration object to provide helper methods for manipulating configs and a container for static methods to work with configs.
+1. `RuleLoader` - a class that aids in loading rules from a directory.
+
+The `@eslint/config` package is intended to add back functionality that would be removed from ESLint with this design.
+
+**Note:** It is not required to use this package with `.eslint.config.js`; this is an optional tool that users may choose to use for convenience.
+
+#### `Config` Class
+
+The `Config` class has this form:
+
+```js
+class Config {
+
+    // pass in any object literal to wrap in Config
+    constructor(object) {}
+
+    // convenience method for easily setting rule severity
+    setRuleSeverity(ruleId, severity) {}
+
+    // determine if an object is a Config
+    static isConfig(object) {}
+
+    // create a new config based on another config
+    static create(baseConfig, ...overrideConfigs) {}
+
+    // merge info from other configs into a base config
+    static assign(receiverConfig, ...supplierConfigs) {}
+}
+```
+
+A new instance of `Config` is created like this:
+
+```js
+const { Config } = require("@eslint/config");
+
+const config = new Config({
+    files: "*.js",
+    ruledefs: {
+        react: require("eslint-plugin-react").rules,
+    },
+    rules: {
+        "react/jsx-uses-react": "error",
+        semi: ["error", "always"]
+    }
+});
+```
+
+After that point, you can easily change the severity of a rule like this:
+
+```js
+config.setRuleSeverity("semi", "warn");
+```
+
+In this way, users don't need to modify an array to change a rule's severity, which is a major downside of this design when using a vanilla object structure as a config.
+
+The static methods on `Config` make it easy to create and modify configs. For example, suppose you want to create a config based on `eslint-config-standard` but would prefer `semi` be a warning instead of an error, you can do so using `Config.create()`:
+
+```js
+const standardConfig = require("eslint-config-standard");
+const { Config } = require("@eslint/config");
+
+const myConfig = Config.create(standardConfig, {
+    rules: {
+        semi: "warn"
+    }
+});
+
+exports.config = myConfig;
+```
+
+Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)
+
+Similarly, the `Config.assign()` method lets users modify their config by merging in changes from other configs, such as:
+
```

Comment:
what happens if merging 2 files/ignores:

```js
var base = { ignores: ["dist/*", "coverage/*", ... ]};

var myConfig = Config.create(base, {ignores: [".*"]});  // ignores = ???
```

Reply:
Right now, `ignore` gets overwritten, which I think is the correct behavior. (You probably don't want a shareable config determining what your project does and doesn't ignore.)

Reply:
You will if it’s not super easy to handle gitignore :-)